### PR TITLE
Update `didEncounterError` implemenation and docs

### PR DESCRIPTION
**Pull request by trevor-scheer**
Sort of follow up to comments from #146.

The changes in https://github.com/apollographql/datasource-rest/pull/107 imposed a new requirement - either an error must be thrown or a response returned (from https://github.com/apollographql/datasource-rest/pull/107/files#diff-b20ea0c217d93b2062f468593a146a7574ae2d53733515f75b950bb584a50f86R353). Given that this error must be thrown, the default implementation of `didEncounterError` is redundant.

Furthermore, this updates the docs based on this less-strict-than-before requirement and clarifies the options available to implementors of this hook.

Within your implementation of `didEncounterError`, you can:
1) observe the error for logging and do nothing else
2) modify the incoming error (which will be thrown immediately after)
3) throw your own error (which bypasses throwing the original error immediately after)

### CODE REVIEW

**Comment by trevor-scheer** [comment]
I think this explicitly doesn't need a changeset since this should be an effective noop.

File: src/RESTDataSource.ts [hunk 1]
```diff
@@ -257,9 +257,7 @@ export abstract class RESTDataSource {
     request: FetcherRequestInit,
   ): ValueOrPromise<CacheOptions | undefined>;
 
-  protected didEncounterError(error: Error, _request: RequestOptions) {
-    throw error;
-  }
+  protected didEncounterError?(error: Error, _request: RequestOptions): void;
```
Comment (glasser) [inline src/RESTDataSource.ts:260]: Hmm. If you implemented didEncounterError by calling `super.didEncounterError`, it's now broken. Maybe leave as empty-bodied?

Comment (trevor-scheer) [reply src/RESTDataSource.ts:260]: Good catch

Comment (trevor-scheer) [reply src/RESTDataSource.ts]: [f4fff03](https://github.com/apollographql/datasource-rest/pull/147/commits/f4fff03e8eff9fbb24926ccb3cd93e2fa3ebb297)

File: src/RESTDataSource.ts [hunk 2]
```diff
@@ -531,7 +534,7 @@ export abstract class RESTDataSource {
             },
           };
         } catch (error) {
-          this.didEncounterError(error as Error, outgoingRequest);
+          this.didEncounterError?.(error as Error, outgoingRequest);
```
Comment (glasser) [inline src/RESTDataSource.ts:537]: can perhaps undo this now?

**Review by glasser — APPROVED** [review]

File: src/RESTDataSource.ts (hunk 2 above)
Comment (trevor-scheer) [reply src/RESTDataSource.ts]: [0a395b4](https://github.com/apollographql/datasource-rest/pull/147/commits/0a395b4a4e28a153e7e5e9f59d7a4d5844b891cb)

**Pull request merged**